src/compiler/binder.ts

```diff
@@ -3333,13 +3333,112 @@ namespace ts {
             }
         }
 
+        function extractJSDocDeclarations<T>(node: T): T[];
+        function extractJSDocDeclarations(node: Node): Node[] {
+            if (!isFunctionDeclaration(node) && !isMethodDeclaration(node)) {
+                return [node];
+            }
+            if (!isInJSFile(node) || !hasJSDocNodes(node)) {
+                return [node];
+            }
+            const declarations: Node[] = [];
+            const jsDocNodes: JSDoc[] = node.jsDoc!;
+            const lastJsDocNode = last(jsDocNodes);
+            let jsDocNodesBuffer: JSDoc[] = [];
+            let currentPos = node.pos;
+            for (const jsDoc of jsDocNodes) {
+                jsDocNodesBuffer.push(jsDoc);
+                if (jsDoc === lastJsDocNode) {
+                    break;
+                }
+                const overloadTag = jsDoc.tags && find(jsDoc.tags, isJSDocOverloadTag);
+                if (!overloadTag) {
+                    continue;
+                }
+                const newParameters: ParameterDeclaration[] = [];
+                for (const tag of jsDoc.tags) {
+                    if (isJSDocParameterTag(tag) && tag.name.kind === SyntaxKind.Identifier) {
+                        const newParameter = factory.createParameterDeclaration(

```

Comment:
Synthesising declarations is not a technique that works in the typescript compiler. A better approach would be to make sure all the jsdoc nodes are still available to the checker. In the checker, make sure that each `@overload` results in a signature. You'll want to start near getSignatureFromDeclaration, or its callers, I think.

Reply:
@sandersn Agreed. I've also observed other BUGs related to IDE integrations that are coming from the fact that I am trying to synthesize those declarations, e.g. "did not expect identifier in trivia".

> You'll want to start near getSignatureFromDeclaration, or its callers, I think.

Thank you for this suggestion! I've already tried that and it seems to be working much better:

https://github.com/microsoft/TypeScript/compare/main...apendua:js-doc-overload-tag?expand=1

How we should go about that? Should I close this pull request and open a new one?

Reply:
It is up to you. I am fine with proceeding with this one; I squash merge PRs since it makes the git history easier to search.

Reply:
@sandersn I ended up opening a new PR here: https://github.com/microsoft/TypeScript/pull/51234

I also used this opportunity to rebase onto the latest `main`. Thank you again for all your suggestions!